httpclient-core/src/main/java/esa/httpclient/core/CompositeRequest.java

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright 2021 OPPO ESA Stack Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package esa.httpclient.core;
+
+import esa.commons.Checks;
+import esa.commons.collection.HashMultiValueMap;
+import esa.commons.collection.MultiValueMap;
+import esa.commons.http.HttpMethod;
+import esa.httpclient.core.netty.NettyHttpClient;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public class CompositeRequest extends HttpRequestBaseImpl implements PlainRequest, FileRequest,
+        MultipartRequest, HttpRequestFacade {
+
+    private static final String DEFAULT_BINARY_CONTENT_TYPE = "application/octet-stream";
+    private static final String DEFAULT_TEXT_CONTENT_TYPE = "text/plain";
+
+    private final NettyHttpClient client;
+    private final Supplier<ChunkRequest> request;
+
+    /**
+     * Body data
+     */
+    private final MultiValueMap<String, String> attributes = new HashMultiValueMap<>();
+    private final List<MultipartFileItem> files = new ArrayList<>();
+    private volatile byte[] bytes;
+    private volatile File file;
+    private volatile boolean multipartEncode = true;
+    private boolean started;
+
+    /**
+     * 1: multipart; 2: segment; 3. file; others: plain
+     */
+    private volatile byte type = 0;
+
+    public CompositeRequest(HttpClientBuilder builder,
+                            NettyHttpClient client,
+                            Supplier<ChunkRequest> request,
+                            HttpMethod method,
+                            String uri) {
+        super(builder, method, uri);
+        Checks.checkNotNull(client, "NettyHttpClient must not be null");
+        Checks.checkNotNull(request, "ChunkRequest must not be null");
+        this.client = client;
+        this.request = request;
+    }
+
+    @Override
+    public synchronized CompletableFuture<HttpResponse> execute() {
+        if (started) {
+            throw new IllegalStateException("The execute() has been called before");
+        }
+        this.started = true;
+        return client.execute(this, ctx, handle, handler);
+    }
+
+    @Override
+    public synchronized PlainRequest body(byte[] bytes) {
```

Comment:
What about using `Buffer` as the parameter

Reply:
done